gadget/mountedfilesystem.go

```diff
@@ -179,10 +198,555 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+	// check how much of the prefix needs to be created
+	for prefix := filepath.Dir(target); prefix != "." && prefix != "/"; prefix = filepath.Dir(prefix) {
+		prefixDst, prefixBackupBase := f.entryDestPaths(dstRoot, "", prefix, backupDir)
+
+		prefixBackupName := prefixBackupBase + ".backup"
+		if osutil.FileExists(prefixBackupName) {
+			continue
+		}
+		if !osutil.IsDirectory(prefixDst) {
+			// does not exist now, will be created on the fly and
+			// removed during rollback
+			continue
+		}
+		if err := os.MkdirAll(filepath.Dir(prefixBackupName), 0755); err != nil {
+			return fmt.Errorf("cannot create backup prefix: %v", err)
+		}
+		if err := makeStamp(prefixBackupName); err != nil {
+			return fmt.Errorf("cannot create a checkpoint for directory: %v", err)
+		}
+	}
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName, dstName, backupBase := f.entryPaths(dstRoot, source, target, backupDir)
+
+	backupName := backupBase + ".backup"
+	sameStamp := backupBase + ".same"
+	preserveStamp := backupBase + ".preserve"
+
+	if !osutil.FileExists(dstName) {
+		// destination does not exist and will be created when writing
+		// the udpate, no need for backup
+		return nil
+	}
+
+	if osutil.FileExists(backupName) || osutil.FileExists(sameStamp) {
+		// file already checked, either has a backup or is the same as
+		// the update, move on
+		return nil
+	}
+
+	if strutil.SortedListContains(preserveInDst, dstName) {
+		// file is to be preserved, create a relevant stamp
+
+		if !osutil.FileExists(dstName) {
+			// preserve, but does not exist, will be written anyway
+			return nil
+		}
+		if osutil.FileExists(preserveStamp) {
+			// already stamped
+			return nil
+		}
+		// make a stamp
+		if err := makeStamp(preserveStamp); err != nil {
+			return fmt.Errorf("cannot create preserve stamp: %v", err)
+		}
+		return nil
+	}
+
+	// try to find out whether the update and the existing file are
+	// identical
+
+	orig, err := os.Open(dstName)
+	if err != nil {
+		return fmt.Errorf("cannot open destination file: %v", err)
+	}
+
+	// backup of the original content
+	backup, err := newStampFile(backupName)
+	if err != nil {
+		return fmt.Errorf("cannot create backup file: %v", err)
+	}
+	// becomes a backup copy or a noop if canceled
+	defer backup.Commit()
+
+	// checksum the original data while it's being copied
+	origHash := crypto.SHA1.New()
+	htr := io.TeeReader(orig, origHash)
+
+	_, err = io.Copy(backup, htr)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot backup original file: %v", err)
+	}
+
+	// digest of the update
+	updateDigest, _, err := osutil.FileDigest(srcName, crypto.SHA1)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot checksum update file: %v", err)
+	}
+	// digest of the currently present data
+	origDigest := origHash.Sum(nil)
+
+	if bytes.Equal(origDigest, updateDigest) {
```

Comment:
Given that we read both files would it be less costly to simply compare block by block and avoid any hash selection and collision vulnerability problems?

Reply:
I think that's a valid concern. I should probably address it in a follow-up though as a similar change will be done in `raw.go`

Reply:
probably worth leaving a TODO then